<a href="https://colab.research.google.com/github/sherjahong1r/Machine-Learning-Lessons/blob/main/02_MLga_tayyorgarlik_NaN_qiymatlar_3_QADAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***3 - QADAM***

# **MLga tayyorgarlik.**

# **NaN** **qiymatlar**

# **NEW SECTION**

In [70]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn # scite-learn kutubxonasi

In [71]:
url = "https://github.com/ageron/handson-ml2/blob/master/datasets/housing/housing.csv?raw=true"
df = pd.read_csv(url)
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [72]:
df.shape

(20640, 10)

Datasetning har bir qatori bu bitta hudud (mahalla, kvartal, blok) haqida ma'lumot. Dataset 10 ta ustundan iborat:

longitude va latitude birgalikda GPS koordinatalar

housing_median_age - uylarning median yoshi

total_rooms - ushbu hududda mavjud jami xonalar

total_bedrooms - ushbu hududda mavjud jami yotoqxonalar

population - hudud aholi soni

households - hududdagi oilalar soni

median_income - median daromad

median_house_values - uylarning median narxi

ocean_proximity - hududning okeanga qanchalik yaqinligi

In [73]:
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(df, test_size=0.2, random_state=50)

In [74]:
train_set

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   
...          ...       ...                 ...          ...             ...   
14565    -117.20     32.84                32.0       2033.0           394.0   
15649    -122.41     37.79                52.0       2161.0           544.0   
10123    -117.92     33.93                12.0       4415.0           890.0   
5600     -118.27     33.79                39.0       1513.0           365.0   
14000    -117.24     34.59                 4.0       5027.0           797.0   

       population  households  median_income  median_house_value  \
4450       1312.0       323.0         1.9375            168800.0   
14253      2775.0       644.0         1.7075             86900.0   
15338      1428.0       407.0         2.4754            157300.0   
18156      1343.0       455.0         6.0777            327500.0   
13431      2231.0       742.0         4.1399            115400.0   
...           ...         ...            ...                 ...   
14565       989.0       389.0         3.2583            181400.0   
15649       904.0       431.0         3.5066            350000.0   
10123      1532.0       854.0         3.7500            166300.0   
5600       1227.0       354.0         3.3929            184600.0   
14000      1869.0       686.0         3.5507            186100.0   

      ocean_proximity  
4450        <1H OCEAN  
14253      NEAR OCEAN  
15338      NEAR OCEAN  
18156       <1H OCEAN  
13431          INLAND  
...               ...  
14565      NEAR OCEAN  
15649        NEAR BAY  
10123       <1H OCEAN  
5600        <1H OCEAN  
14000          INLAND  

[16512 rows x 10 columns]

In [75]:
test_set

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
1945     -120.93     38.77                 9.0       2229.0           355.0   
18006    -121.97     37.30                31.0       3340.0           735.0   
13062    -121.32     38.57                15.0       3369.0           499.0   
13396    -117.57     34.02                 5.0       6933.0          1311.0   
9787     -120.79     36.06                29.0       1916.0           386.0   
...          ...       ...                 ...          ...             ...   
15490    -117.19     33.14                12.0       3652.0           923.0   
4339     -118.29     34.08                34.0        479.0           182.0   
14248    -117.10     32.72                 5.0       1615.0           387.0   
9060     -118.14     34.68                25.0       1703.0           342.0   
12959    -121.30     38.71                17.0       5434.0          1106.0   

       population  households  median_income  median_house_value  \
1945        788.0       341.0         5.5111            196300.0   
18006      1891.0       686.0         4.8542            275000.0   
13062      1733.0       470.0         5.3100            127500.0   
13396      3845.0      1285.0         4.6727            158900.0   
9787       1019.0       314.0         2.4881             87500.0   
...           ...         ...            ...                 ...   
15490      1677.0       728.0         2.3267             92000.0   
4339        557.0       170.0         1.5250            210000.0   
14248      1094.0       394.0         2.2024            137200.0   
9060        775.0       309.0         4.5455            126500.0   
12959      2755.0      1047.0         2.8226             99900.0   

      ocean_proximity  
1945           INLAND  
18006       <1H OCEAN  
13062          INLAND  
13396          INLAND  
9787        <1H OCEAN  
...               ...  
15490       <1H OCEAN  
4339        <1H OCEAN  
14248      NEAR OCEAN  
9060           INLAND  
12959          INLAND  

[4128 rows x 10 columns]

In [76]:
housing = train_set.drop("median_house_value", axis=1)
# df dan median_house_value ustunini o'chirib qolgan parametr yani ustunlarni housing ga saqladik
housing_labels = train_set["median_house_value"].copy()
# median_house_value ni o'zini housing_labels ga saqladik

In [77]:
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   

       population  households  median_income ocean_proximity  
4450       1312.0       323.0         1.9375       <1H OCEAN  
14253      2775.0       644.0         1.7075      NEAR OCEAN  
15338      1428.0       407.0         2.4754      NEAR OCEAN  
18156      1343.0       455.0         6.0777       <1H OCEAN  
13431      2231.0       742.0         4.1399          INLAND

In [78]:
housing.shape

(16512, 9)

In [79]:
housing_labels.head()

4450     168800.0
14253     86900.0
15338    157300.0
18156    327500.0
13431    115400.0
Name: median_house_value, dtype: float64

In [80]:
housing_labels.shape

(16512,)

In [81]:
df.info()
# total_bedrooms to'liq emas ekan yani NaN qiymatlar mavjud ekan

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [82]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16512 entries, 4450 to 14000
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           16512 non-null  float64
 1   latitude            16512 non-null  float64
 2   housing_median_age  16512 non-null  float64
 3   total_rooms         16512 non-null  float64
 4   total_bedrooms      16351 non-null  float64
 5   population          16512 non-null  float64
 6   households          16512 non-null  float64
 7   median_income       16512 non-null  float64
 8   ocean_proximity     16512 non-null  object 
dtypes: float64(8), object(1)
memory usage: 1.3+ MB


In [83]:
16512-16351

161

## NaN bo'lgan ustun muammolarini hal qilish uchun bir nechta usullardan foydalanish mumkin

## 1-Usul NaN bo'lgan qatorlarni tashlab yuborish

In [84]:
housing.dropna(subset=['total_bedrooms'])
# housing ning qatorlari 16512 ta edi total_bedrooms NaN larini o'chirgandan so'ng 16351 ta qoldi

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   
...          ...       ...                 ...          ...             ...   
14565    -117.20     32.84                32.0       2033.0           394.0   
15649    -122.41     37.79                52.0       2161.0           544.0   
10123    -117.92     33.93                12.0       4415.0           890.0   
5600     -118.27     33.79                39.0       1513.0           365.0   
14000    -117.24     34.59                 4.0       5027.0           797.0   

       population  households  median_income ocean_proximity  
4450       1312.0       323.0         1.9375       <1H OCEAN  
14253      2775.0       644.0         1.7075      NEAR OCEAN  
15338      1428.0       407.0         2.4754      NEAR OCEAN  
18156      1343.0       455.0         6.0777       <1H OCEAN  
13431      2231.0       742.0         4.1399          INLAND  
...           ...         ...            ...             ...  
14565       989.0       389.0         3.2583      NEAR OCEAN  
15649       904.0       431.0         3.5066        NEAR BAY  
10123      1532.0       854.0         3.7500       <1H OCEAN  
5600       1227.0       354.0         3.3929       <1H OCEAN  
14000      1869.0       686.0         3.5507          INLAND  

[16351 rows x 9 columns]

## 2-Usul. Ustunni tashlab yuborish

In [85]:
housing.drop('total_bedrooms', axis=1)

longitude  latitude  housing_median_age  total_rooms  population  \
4450     -118.21     34.07                42.0        902.0      1312.0   
14253    -117.10     32.71                29.0       3422.0      2775.0   
15338    -117.29     33.20                16.0       2150.0      1428.0   
18156    -122.06     37.36                35.0       2693.0      1343.0   
13431    -117.42     34.10                18.0       3977.0      2231.0   
...          ...       ...                 ...          ...         ...   
14565    -117.20     32.84                32.0       2033.0       989.0   
15649    -122.41     37.79                52.0       2161.0       904.0   
10123    -117.92     33.93                12.0       4415.0      1532.0   
5600     -118.27     33.79                39.0       1513.0      1227.0   
14000    -117.24     34.59                 4.0       5027.0      1869.0   

       households  median_income ocean_proximity  
4450        323.0         1.9375       <1H OCEAN  
14253       644.0         1.7075      NEAR OCEAN  
15338       407.0         2.4754      NEAR OCEAN  
18156       455.0         6.0777       <1H OCEAN  
13431       742.0         4.1399          INLAND  
...           ...            ...             ...  
14565       389.0         3.2583      NEAR OCEAN  
15649       431.0         3.5066        NEAR BAY  
10123       854.0         3.7500       <1H OCEAN  
5600        354.0         3.3929       <1H OCEAN  
14000       686.0         3.5507          INLAND  

[16512 rows x 8 columns]

## 3-Usul. NaN qiymatlarni boshqa qiymatlar bilan almashtirish


In [86]:
median = housing['total_bedrooms'].median()
median

436.0

In [87]:
housing['total_bedrooms'].fillna(median)

4450     318.0
14253    713.0
15338    461.0
18156    493.0
13431    809.0
         ...  
14565    394.0
15649    544.0
10123    890.0
5600     365.0
14000    797.0
Name: total_bedrooms, Length: 16512, dtype: float64

## New Section

## Simplelmputer

In [88]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy="median")

# sklearn.impute modulidan SimpleImputer klassi import qilinadi. So'ngra, ushbu klassdan
# imputer nomli obyekt yaratiladi. strategy="median" parametri esa, ma'lumotlardagi
# NaN (yetishmayotgan) qiymatlarni ustunning medianasi bilan to'ldirish uchun ishlatilishini bildiradi.

In [89]:
housing_num = housing.drop("ocean_proximity", axis=1)
# ocean_proximity degan string ustunini o'chirib housing_num ga saqladik

In [90]:
imputer.fit(housing_num)
# Ushbu kod yetishmayotgan qiymatlarni (NaN) to'ldirish uchun medianani hisoblaydi va imputer obyektida saqlaydi.
# fit() metodi housing_num ma'lumotlaridan median qiymatlarni hisoblab, ularni ichki holatida saqlaydi.
# Agar bunga housing ni kiritganimizda hato chiqarardi chunki matnli ustun bor
# Shuning uchun uni housing_num o'zgaruvchisida o'chirib housing_num ni fit ga uzatayabmiz

SimpleImputer(strategy='median')

In [91]:
imputer.statistics_
# Bu metod har bir ustun uchun medianni hisoblaydi tartibli ravishda.

array([-118.5   ,   34.26  ,   29.    , 2131.    ,  436.    , 1167.    ,
        410.    ,    3.5299])

In [92]:
X = imputer.transform(housing_num)
X
# imputer.transform(housing_num) kodi housing_num ma'lumotlaridagi bo'sh (NaN)
# qiymatlarni avval hisoblangan medianalar bilan to'ldirib va X ga saqlanayabdi

array([[-1.1821e+02,  3.4070e+01,  4.2000e+01, ...,  1.3120e+03,
         3.2300e+02,  1.9375e+00],
       [-1.1710e+02,  3.2710e+01,  2.9000e+01, ...,  2.7750e+03,
         6.4400e+02,  1.7075e+00],
       [-1.1729e+02,  3.3200e+01,  1.6000e+01, ...,  1.4280e+03,
         4.0700e+02,  2.4754e+00],
       ...,
       [-1.1792e+02,  3.3930e+01,  1.2000e+01, ...,  1.5320e+03,
         8.5400e+02,  3.7500e+00],
       [-1.1827e+02,  3.3790e+01,  3.9000e+01, ...,  1.2270e+03,
         3.5400e+02,  3.3929e+00],
       [-1.1724e+02,  3.4590e+01,  4.0000e+00, ...,  1.8690e+03,
         6.8600e+02,  3.5507e+00]])

In [93]:
type(X)

numpy.ndarray

In [94]:
X = imputer.fit_transform(housing_num)
X
# Bu yuqoridagi kodlarni alohida qilib yozgan edik bir qatorga yozsa ham bo'ladi

array([[-1.1821e+02,  3.4070e+01,  4.2000e+01, ...,  1.3120e+03,
         3.2300e+02,  1.9375e+00],
       [-1.1710e+02,  3.2710e+01,  2.9000e+01, ...,  2.7750e+03,
         6.4400e+02,  1.7075e+00],
       [-1.1729e+02,  3.3200e+01,  1.6000e+01, ...,  1.4280e+03,
         4.0700e+02,  2.4754e+00],
       ...,
       [-1.1792e+02,  3.3930e+01,  1.2000e+01, ...,  1.5320e+03,
         8.5400e+02,  3.7500e+00],
       [-1.1827e+02,  3.3790e+01,  3.9000e+01, ...,  1.2270e+03,
         3.5400e+02,  3.3929e+00],
       [-1.1724e+02,  3.4590e+01,  4.0000e+00, ...,  1.8690e+03,
         6.8600e+02,  3.5507e+00]])

In [95]:
housing_tr = pd.DataFrame(X, columns=housing_num.columns, index=housing_num.index)
housing_tr
# Ushbu kod X deb nomlangan NumPy massivini housing_num DataFrame'ining ustun nomlari
# va indekslarini saqlagan holda yangi housing_tr nomli Pandas DataFrame'ga aylantiradi.
# tr - transform degani

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   
...          ...       ...                 ...          ...             ...   
14565    -117.20     32.84                32.0       2033.0           394.0   
15649    -122.41     37.79                52.0       2161.0           544.0   
10123    -117.92     33.93                12.0       4415.0           890.0   
5600     -118.27     33.79                39.0       1513.0           365.0   
14000    -117.24     34.59                 4.0       5027.0           797.0   

       population  households  median_income  
4450       1312.0       323.0         1.9375  
14253      2775.0       644.0         1.7075  
15338      1428.0       407.0         2.4754  
18156      1343.0       455.0         6.0777  
13431      2231.0       742.0         4.1399  
...           ...         ...            ...  
14565       989.0       389.0         3.2583  
15649       904.0       431.0         3.5066  
10123      1532.0       854.0         3.7500  
5600       1227.0       354.0         3.3929  
14000      1869.0       686.0         3.5507  

[16512 rows x 8 columns]

## Umumiy maqsad yani NaN bo'lgan qatorlar o'rnini to'ldirish edi biz uni median bilan to'ldirdik

# **NEW SECTION**

# **Matnlar. SimpleEncoder**

## Maqsad matnli ustunni sonli ustunga aylantirish

> Add blockquote



In [96]:
housing_cat = housing[['ocean_proximity']]
housing_cat

ocean_proximity
4450        <1H OCEAN
14253      NEAR OCEAN
15338      NEAR OCEAN
18156       <1H OCEAN
13431          INLAND
...               ...
14565      NEAR OCEAN
15649        NEAR BAY
10123       <1H OCEAN
5600        <1H OCEAN
14000          INLAND

[16512 rows x 1 columns]

In [97]:
housing_cat.value_counts()

ocean_proximity
<1H OCEAN          7255
INLAND             5286
NEAR OCEAN         2142
NEAR BAY           1825
ISLAND                4
Name: count, dtype: int64

In [98]:
from sklearn.preprocessing import OrdinalEncoder
ordinal_encoder = OrdinalEncoder()
housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat)
housing_cat_encoded
# from sklearn.preprocessing import OrdinalEncoder: Bu kod sklearn kutubxonasidan OrdinalEncoder sinfini import
# qiladi. OrdinalEncoder matnli (kategorik) ma'lumotlarni tartiblangan sonli qiymatlarga o'tkazish uchun ishlatiladi.

# ordinal_encoder = OrdinalEncoder(): Bu yerda OrdinalEncoder sinfining obyekti yaratiladi. Bu obyekt matnli
# ma'lumotlarni qanday o'zgartirishni o'rganish va o'zgartirish uchun ishlatiladi.

# housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat): Bu qator ikkita asosiy ishni bajaradi:

# fit(housing_cat): ordinal_encoder obyekti housing_cat DataFrame'idagi noyob kategoriyalarni (masalan,
# '<1H OCEAN', 'NEAR OCEAN', 'INLAND' va boshqalar) aniqlaydi va ularga sonli qiymatlarni (0, 1, 2, ...) belgilaydi.
# transform(housing_cat): Keyin, housing_cat DataFrame'idagi har bir kategorik qiymatni o'rgangan sonli qiymatlariga
# almashtiradi. Natijada, ocean_proximity ustunidagi matnli qiymatlar sonlarga aylantiriladi
# va housing_cat_encoded nomli yangi NumPy massiviga saqlanadi.

# OrdinalEncoder ocean_proximity ustunidagi 5 ta noyob kategoriyani 0 dan 4 gacha bo'lgan
# sonli qiymatlarga aylantiradi. Har bir noyob kategoriya o'zining noyob sonli kodiga ega bo'ladi.

array([[0.],
       [4.],
       [4.],
       ...,
       [0.],
       [0.],
       [1.]])

In [99]:
housing_cat_encoded[:10]
# Bu bilan 10 gacha bo'lganlarni chiqaradi

array([[0.],
       [4.],
       [4.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.]])

In [100]:
housing_cat_encoded

array([[0.],
       [4.],
       [4.],
       ...,
       [0.],
       [0.],
       [1.]])

Bu yuqoridagi codning kamchilik tomoni:

ML bir-biriga yaqinroq bo'lgan sonlarni yaqinroq deb o'ylaydi yani 0 va 1 ni
lekin 0 yani okeanga 1 soatlik yo'l bilan, 1 yani quruqlik ichkarisidagining ancha farqi bor shu sababli hatolik ketishi mumkin:

Bu yerda esa 0 va 4 qiymatlari yaqinroq, lekin ML bularni bir biridan eng uzoq qiymat deb tahlil qiladi.

0 = 1H OCEAN yani okenga 1 soat lik yo'l degan

1 = INLAND yani quruqlikning ancha ichkari qismida degani

2 = ISLAND orolda joylashgan degani

3 = NEARBY yaqinroq degani

4 = NEAR OKEAN  yani okeanga yaqin degani



In [101]:
housing.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   
1933     -120.99     38.69                 5.0       5743.0          1074.0   
8437     -118.37     33.93                46.0       1130.0           201.0   
10944    -117.87     33.74                16.0       1243.0           365.0   
7646     -118.27     33.81                42.0        865.0           208.0   
11047    -117.83     33.81                28.0       1972.0           315.0   

       population  households  median_income ocean_proximity  
4450       1312.0       323.0         1.9375       <1H OCEAN  
14253      2775.0       644.0         1.7075      NEAR OCEAN  
15338      1428.0       407.0         2.4754      NEAR OCEAN  
18156      1343.0       455.0         6.0777       <1H OCEAN  
13431      2231.0       742.0         4.1399          INLAND  
1933       2651.0       962.0         4.1163          INLAND  
8437        503.0       196.0         4.4861       <1H OCEAN  
10944      1925.0       376.0         2.7632       <1H OCEAN  
7646        811.0       218.0         3.8621       <1H OCEAN  
11047       970.0       326.0         5.4298       <1H OCEAN

## Shuning uchun boshqa metoddan foydalanamiz

# **NEW SECTION**

In [102]:
from sklearn.preprocessing import OneHotEncoder

cat_encoder = OneHotEncoder()
housing_cat_1hot = cat_encoder.fit_transform(housing_cat)
housing_cat_1hot.toarray()

# from sklearn.preprocessing import OneHotEncoder: Bu sklearn kutubxonasidan OneHotEncoder
# sinfini import qiladi. OneHotEncoder matnli kategoriyalarni har bir kategoriya
# uchun alohida ikkilik (0 yoki 1) ustunlarga aylantirish uchun ishlatiladi.

# cat_encoder = OneHotEncoder(): Bu yerda OneHotEncoder obyekti yaratiladi. Bu obyekt kategoriyalarni kodlashni o'rganadi.

# housing_cat_1hot = cat_encoder.fit_transform(housing_cat): Bu qator ikkita asosiy ishni bajaradi:

# fit(housing_cat): cat_encoder obyekti housing_cat DataFrame'idagi noyob
# kategoriyalarni (masalan, '<1H OCEAN', 'NEAR OCEAN' va hokazo) aniqlaydi.
# transform(housing_cat): Keyin, har bir kategorik qiymatni One-Hot encoded massivga aylantiradi.
# Ya'ni, har bir noyob kategoriya uchun yangi ustun yaratiladi va agar qatordagi qiymat shu kategoriyaga
# to'g'ri kelsa, ustunda 1, aks holda 0 bo'ladi. Natijada, housing_cat_1hot nomli yangi scipy sparse matrix (siyrak matritsa) saqlanadi.
# housing_cat_1hot.toarray(): fit_transform natijasi siyrak matritsa ko'rinishida bo'lgani sababli, toarray()
# metodi uni to'liq NumPy massiviga aylantiradi, bu esa natijani ko'rish va ishlash uchun qulayroqdir.



array([[1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 1.],
       ...,
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.]])

In [103]:
housing_cat[:5]

ocean_proximity
4450        <1H OCEAN
14253      NEAR OCEAN
15338      NEAR OCEAN
18156       <1H OCEAN
13431          INLAND

OneHotEncoder qisqacha vazifasi:

     1,       0,         0,          0        0

<1H OCEAN,  INLAND,  NEAR OCEAN,  NEAR BAY,  ISLAND

Yani har bir 5 ta katagoriya yani qiymat uchun 5 ta ustun yaratadi agar shu qiymat bor bo'lsa 1 aks holda 0 qo'yiladi



## Bu kodning kamchilik tomoni agar categoriyalar ko'p bo'lsa buning ishlashi sekinlashadi bizda kamroq 5 kategoriya bo'lgani uchun mos tushadi.

## Yuqoridagi barcha kodlardan NaN codlarni alamashtirish, ocean_proximity ustuni qiymatlarini sonlarga o'girishda biz transformerlardan foydalanganmiz yani fit.transform kabilardan.

### 3.3 Transfomer

Yuqorida ma'lumotlarg ishlov berishda biz `sklearn` tarkibidagi tayyor obyektlar va ulardagi `.fit()` yoki `.fit_transform()` funksiyalariga murojaat qildik.
Bu funksiyalar (`imputer`, `OneHotEncoder` va hokazo) **transformer**lar deb ataladi.

Umuman olganda `sklearn` tarkibidagi obyektlar uch turli bo'ladi:
1. **Estimator** - tahmin qiluvchi (baholovchi) funksiyalar. Bunday obyektlar `.fit()` metodiga ega bo'ladi, bu metod berilgan ma'lumotlardan qandaydur qiymatni hisoblaydi (masalan, avvalroq biz `imputer.fit()` yordamida median qiymatni hisobladik)
2. **Transformer** - ham tahmin qilib ham ma'lumotga ishlov beruvchi obyekt. Bunday obyektlar `.fit_transform()` metodiga ega bo'ladi, ya'ni avval qandaydur qiymatnbi hisoblaydi, keyin esa hisoblashni ma'lumotlarga qo'llaydi. Yuoqirdagi `OrdinalEncoder` va `OneHotEncoder` shunday obyektlardan.
3. **Predictor** - bashorat qiluvchi obyektlar. Bunday obyektlar `.predict()` metodiga ega bo'ladi. Ular biz bergan ma'lumotlar asosida bashoratlar qaytaradi. Dars davomida predictor obyektlar bilan ko'p ishlaymiz.

Obyektlarning afzalligi, ular hisob kitob natijalarini ichki attributrlarda saqlab qoladi. Masalan, yuqorida ko'rganimiz `imputer.statistics_`.

Keling endi o'zimiz ham transformer yozishni o'rganamiz.

Esingizda bo'lsa, avvalgi darsimizda biz ma'lumotlarg yangi 2 ustun qo'shdik: `rooms_per_household` va `bedrooms_per_room`.

Keling berilgan ma'lumotlarga shu ikki ustunni avtomat ravishda qo'shadigan transformer yasaymiz.

Buning uchun `sklearn` tarkibida `BaseEstimator` va `TransformerMixin` klasslaridan meros olib yangi klass yasaymiz va bu klassimizga kabi `fit()` va `transform()` metodlarini qo'shamiz:

# **NEW SECTION**

# **Transformer yasash**

In [104]:
df.head(1)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY

In [105]:
from sklearn.base import BaseEstimator, TransformerMixin

# bizga kerak ustunlar indexlari
rooms_ix, bedrooms_ix, population_ix, households_ix = 3, 4, 5, 6

class CombinedAttributesadder(BaseEstimator, TransformerMixin):
  def __init__(self, add_bedrooms_per_room=True):
    self.add_bedrooms_per_room = add_bedrooms_per_room
  def fit(self, X, y=None):
    return self # bizning funksiyamiz faqat transformer. estimator emas
  def transform(self, X):
    rooms_per_household = X[:, rooms_ix] / X[:, households_ix]
    population_per_household = X[:, population_ix] / X[:, households_ix]
    if self.add_bedrooms_per_room:  # add_bedrooms_per_room ustuni ixtiyoriy bo'ladi
      bedrooms_per_room = X[:, bedrooms_ix] / X[:, rooms_ix]
      return np.c_[X, rooms_per_household, population_per_household, bedrooms_per_room]
    else:
      return np.c_[X, rooms_per_household, population_per_household]

# from sklearn.base import BaseEstimator, TransformerMixin: Bu qator scikit-learn kutubxonasidan zarur bo'lgan bazaviy
# sinflarni import qiladi. BaseEstimator get_params() va set_params() kabi metodlarni taqdim etadi, TransformerMixin esa
# fit_transform() metodini (u ichkaridan fit() va keyin transform() metodlarini chaqiradi) taqdim etadi.

# rooms_ix, bedrooms_ix, population_ix, households_ix = 3, 4, 5, 6: Bu o'zgaruvchilar kirish ma'lumotlaridagi (X) ma'lum
# ustunlarning indeks raqamlarini saqlaydi. Bu transformerga ushbu ustunlarga raqamli indekslari orqali kirish imkonini beradi.

# class CombinedAttributesadder(BaseEstimator, TransformerMixin):: Bu yerda BaseEstimator va TransformerMixin dan meros
# olgan holda maxsus transformer sinfi aniqlanadi. Bu uning scikit-learn ning pipeline xususiyatlari bilan moslashishini ta'minlaydi.

# def __init__(self, add_bedrooms_per_room=True):: Bu konstruktor. U transformerni ishga tushiradi va qo'shimcha xususiyat
# (bedrooms_per_room) qo'shilishi kerakligini belgilash imkonini beradi. Odatiy holda, u True ga o'rnatilgan.

# def fit(self, X, y=None):: Transformer uchun fit metodi odatda o'quv ma'lumotlaridan parametrlarni o'rganadi. Ushbu aniq
# holatda, bu transformer hech qanday parametrni o'rganishi shart emas, shuning uchun u shunchaki self ni qaytaradi.

# def transform(self, X):: Bu yangi xususiyatlar hisoblanadigan va kirish ma'lumotlariga (X) qo'shiladigan asosiy metod. U quyidagi
# hisob-kitoblarni amalga oshiradi:

# rooms_per_household = X[:, rooms_ix] / X[:, households_ix]: Har bir uy xo'jaligidagi xonalar sonini hisoblaydi.
# population_per_household = X[:, population_ix] / X[:, households_ix]: Har bir uy xo'jaligidagi aholi sonini hisoblaydi.
# if self.add_bedrooms_per_room:: Agar add_bedrooms_per_room True bo'lsa (sukut bo'yicha):
# bedrooms_per_room = X[:, bedrooms_ix] / X[:, rooms_ix]: Har bir xonadagi yotoqxonalar sonini hisoblaydi.
# return np.c_[X, rooms_per_household, population_per_household, bedrooms_per_room]: U asl ma'lumotlarni
# X ni uchta yangi hisoblangan xususiyat bilan birlashtiradi.
# else:: Agar add_bedrooms_per_room False bo'lsa:
# return np.c_[X, rooms_per_household, population_per_household]: U asl ma'lumotlarni
# X ni faqat ikkita yangi hisoblangan xususiyat bilan birlashtiradi.

In [106]:
atter_adder = CombinedAttributesadder()
housing_extra_attribs = atter_adder.transform(housing.values)

# atter_adder = CombinedAttributesadder(): Bu qator CombinedAttributesadder sinfidan atter_adder nomli obyekt yaratadi.
# Bu obyekt ma'lumotlarga yangi xususiyatlarni qo'shish uchun ishlatiladi (masalan, har bir uy xo'jaligidagi xonalar soni kabi).

# housing_extra_attribs = atter_adder.transform(housing.values): Bu qator housing DataFrame'ining qiymatlarini
# (.values orqali NumPy massiviga aylantiriladi) atter_adder obyektiga uzatadi. transform metodi
# yordamida housing.values ma'lumotlariga yangi xususiyatlar qo'shiladi va natija housing_extra_attribs
# o'zgaruvchisiga saqlanadi. Bu yerda housing.values ishlatilishi, transformer matnli ustunlarni kutmasligini bildiradi.

In [107]:
housing.values

array([[-118.21, 34.07, 42.0, ..., 323.0, 1.9375, '<1H OCEAN'],
       [-117.1, 32.71, 29.0, ..., 644.0, 1.7075, 'NEAR OCEAN'],
       [-117.29, 33.2, 16.0, ..., 407.0, 2.4754, 'NEAR OCEAN'],
       ...,
       [-117.92, 33.93, 12.0, ..., 854.0, 3.75, '<1H OCEAN'],
       [-118.27, 33.79, 39.0, ..., 354.0, 3.3929, '<1H OCEAN'],
       [-117.24, 34.59, 4.0, ..., 686.0, 3.5507, 'INLAND']], dtype=object)

In [108]:
housing.values[0,:]

array([-118.21, 34.07, 42.0, 902.0, 318.0, 1312.0, 323.0, 1.9375,
       '<1H OCEAN'], dtype=object)

In [109]:
housing_extra_attribs[0,:]

array([-118.21, 34.07, 42.0, 902.0, 318.0, 1312.0, 323.0, 1.9375,
       '<1H OCEAN', 2.7925696594427243, 4.061919504643963,
       0.352549889135255], dtype=object)

# **NEW SECTION**

# **Normalizatsiya (MinMaxScaler)**

In [110]:
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   

       population  households  median_income ocean_proximity  
4450       1312.0       323.0         1.9375       <1H OCEAN  
14253      2775.0       644.0         1.7075      NEAR OCEAN  
15338      1428.0       407.0         2.4754      NEAR OCEAN  
18156      1343.0       455.0         6.0777       <1H OCEAN  
13431      2231.0       742.0         4.1399          INLAND

In [111]:
housing.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  16512.000000  16512.000000        16512.000000  16512.000000   
mean    -119.568927     35.634159           28.593447   2626.877180   
std        2.006024      2.139083           12.587014   2137.144801   
min     -124.350000     32.540000            1.000000      6.000000   
25%     -121.790000     33.930000           18.000000   1451.000000   
50%     -118.500000     34.260000           29.000000   2131.000000   
75%     -118.010000     37.720000           37.000000   3150.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  
count    16351.000000  16512.000000  16512.000000   16512.000000  
mean       536.276313   1420.163094    497.961846       3.867109  
std        414.875757   1121.241422    375.284263       1.905121  
min          1.000000      3.000000      1.000000       0.499900  
25%        296.000000    788.000000    281.000000       2.562500  
50%        436.000000   1167.000000    410.000000       3.529900  
75%        646.000000   1724.000000    604.000000       4.733225  
max       6445.000000  35682.000000   6082.000000      15.000100

## Etibor bersak har bir ustundagi min va max qiymatlar hammasida har xil masalan housing_median_age ustunida min 1 dan, max 52 gacha yoki total_rooms ustunida min 6 dan, max 39320 gacha.

## Huddi shunday har xil oraliqdagi ustun qiymatlari ML xato ishlashiga olib kelishi mumkin shuning uchun biz bu barcha qiymatlarni 0 va 1 oralig'idagi qiymatlarga almashtiramiz.

## Buning uchun maxsus formula bor

#### 3.4.1 Min-max scaling

Min-max scaling berilgan qiymatlarni 0-1 oralig'iga olib keladi.

Buning uchun biror ustundagi barcha qiymatlarning min ($x_{min}$) va max ($x_{max}$) qiymatlari topiladi. Keyin esa ustundagi har bir qiymat ($x$) quyidagi formula asosida hisoblab chiqiladi:

![](https://arshpreetsingh.files.wordpress.com/2017/03/normal.png)

## Bu pastda formulasiz holda ishlaymiz

In [112]:
xmin = housing['median_income'].min()
xmin

0.4999

In [113]:
xmax = housing['median_income'].max()
xmax

15.0001

In [114]:
xnew = (15-xmin)/(xmax-xmin)
xnew

0.9999931035433994

In [115]:
xnew = (1-xmin)/(xmax-xmin)
xnew

0.034489179459593663

In [116]:
housing_num

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   
...          ...       ...                 ...          ...             ...   
14565    -117.20     32.84                32.0       2033.0           394.0   
15649    -122.41     37.79                52.0       2161.0           544.0   
10123    -117.92     33.93                12.0       4415.0           890.0   
5600     -118.27     33.79                39.0       1513.0           365.0   
14000    -117.24     34.59                 4.0       5027.0           797.0   

       population  households  median_income  
4450       1312.0       323.0         1.9375  
14253      2775.0       644.0         1.7075  
15338      1428.0       407.0         2.4754  
18156      1343.0       455.0         6.0777  
13431      2231.0       742.0         4.1399  
...           ...         ...            ...  
14565       989.0       389.0         3.2583  
15649       904.0       431.0         3.5066  
10123      1532.0       854.0         3.7500  
5600       1227.0       354.0         3.3929  
14000      1869.0       686.0         3.5507  

[16512 rows x 8 columns]

## Maxsus kutubxona yordamida hisoblash

## Bu bilan barcha qiymatlar 0 va 1 oralig'ida joylashadi

In [117]:
from sklearn.preprocessing import MinMaxScaler

min_max_scaler = MinMaxScaler()

min_max_scaler.fit_transform(housing_num)

# from sklearn.preprocessing import MinMaxScaler: Bu scikit-learn kutubxonasidan MinMaxScaler sinfini import qiladi.
# MinMaxScaler ma'lumotlaringizdagi barcha raqamli qiymatlarni aniq bir oraliqqa (odatda 0 dan 1 gacha) keltirish uchun ishlatiladi.

# min_max_scaler = MinMaxScaler(): Bu MinMaxScaler sinfining obyekti yaratadi.
# Bu obyekt ma'lumotlaringizni masshtablash qoidalarini o'rganadi va keyinchalik ularni qo'llaydi.

# min_max_scaler.fit_transform(housing_num): Bu qator ikkita asosiy ishni bajaradi:

# fit(housing_num): housing_num ma'lumotlaridagi har bir ustun uchun minimal va maksimal qiymatlarni hisoblaydi.
# Bu qiymatlar masshtablash formulasi uchun asos bo'ladi.
# transform(housing_num): Keyin, hisoblangan minimal va maksimal qiymatlar yordamida housing_num dagi har
# bir qiymatni 0 va 1 oralig'iga masshtablaydi. Natijada, barcha raqamli ma'lumotlaringiz bir xil masshtabda bo'ladi.

array([[0.61155378, 0.16259299, 0.80392157, ..., 0.03668825, 0.05295182,
        0.09914346],
       [0.72211155, 0.01806589, 0.54901961, ..., 0.07769276, 0.10573919,
        0.08328161],
       [0.70318725, 0.07013815, 0.29411765, ..., 0.03993946, 0.06676533,
        0.1362395 ],
       ...,
       [0.64043825, 0.1477152 , 0.21568627, ..., 0.04285434, 0.14027298,
        0.22414174],
       [0.60557769, 0.13283741, 0.74509804, ..., 0.03430589, 0.05804966,
        0.19951449],
       [0.70816733, 0.21785335, 0.05882353, ..., 0.05229967, 0.11264595,
        0.2103971 ]])

## Hammasi 0 va 1 oralig'iga tushdi

# **NEW SECTION**

# **Standartizatsiya (StandartScaler)**

## Buga ham formula yozsa ham bo'ladi yoki maxusus kutubxona bilan ishlasak ham bo'ladi

#### 3.4.2 Standard Scaler - standartlash
StandardScaler biroz boshqa usulda ishlaydi. Sonlarni standartlas hcuhn bizga berilgan ustun bo'yicha o'rta qiymat ($\mu$) va standart cheklanish ($\sigma$) talab qilinadi. Har bir qatordag qiymat esa quydagicha hisoblanadi

![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQxeMnm1BSa8qyNkCB6wqiYa790XtNxCdM48k_be-9zESUvivv_nuY6SXk8fpJK_psGiA&usqp=CAU)

## Bu kutubxona yordamida 0 va 1 oralig'iga emas balki bir biriga yaqin bo'lgan diapazonga keladi yani tafovuti katta bo'lmaydi

In [118]:
from sklearn.preprocessing import StandardScaler

standard_scaler = StandardScaler()

standard_scaler.fit_transform(housing_num)

# from sklearn.preprocessing import StandardScaler: Bu scikit-learn kutubxonasidan StandardScaler sinfini import qiladi.
# StandardScaler ma'lumotlaringizni o'rtacha qiymati 0 va standart og'ishi 1 bo'lgan taqsimotga aylantiradi.
# Bu ma'lumotlarning tarqalishini bir xil diapazonga keltirishga yordam beradi, lekin 0 va 1 oralig'iga emas,
# balki bir-biriga yaqin bo'lgan diapazonga keladi, ya'ni tafovuti katta bo'lmaydi.

# standard_scaler = StandardScaler(): Bu StandardScaler sinfining obyekti yaratadi. Bu obyekt ma'lumotlaringizni
# standartlashtirish qoidalarini o'rganadi va keyinchalik ularni qo'llaydi.

# standard_scaler.fit_transform(housing_num): Bu qator ikkita asosiy ishni bajaradi:

# fit(housing_num): housing_num ma'lumotlaridagi har bir ustun uchun o'rtacha qiymat (mean) va standart
# og'ish (standard deviation) ni hisoblaydi. Bu qiymatlar standartlashtirish formulasi uchun asos bo'ladi.
# transform(housing_num): Keyin, hisoblangan o'rtacha va standart og'ish yordamida housing_num dagi
# har bir qiymatni standartlashtiradi. Natijada, barcha raqamli ma'lumotlaringiz bir xil masshtabda bo'ladi,
# lekin MinMaxScaler kabi 0-1 oralig'ida emas, balki o'rtacha qiymati 0 va standart og'ishi 1 atrofida bo'ladi.


array([[ 0.67744368, -0.73125093,  1.06514215, ..., -0.09647019,
        -0.46622564, -1.01288451],
       [ 1.23079392, -1.36705675,  0.03230036, ...,  1.20837307,
         0.38915188, -1.13361541],
       [ 1.13607631, -1.13797965, -1.00054142, ...,  0.0069897 ,
        -0.24238853, -0.73053165],
       ...,
       [ 0.82201266, -0.79670153, -1.3183389 , ...,  0.09974684,
         0.94874465, -0.06147245],
       [ 0.64753286, -0.86215213,  0.82679405, ..., -0.17228131,
        -0.38361908, -0.24892031],
       [ 1.16100199, -0.4881487 , -1.95393385, ...,  0.40031566,
         0.50107043, -0.1660884 ]])

In [119]:
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   

       population  households  median_income ocean_proximity  
4450       1312.0       323.0         1.9375       <1H OCEAN  
14253      2775.0       644.0         1.7075      NEAR OCEAN  
15338      1428.0       407.0         2.4754      NEAR OCEAN  
18156      1343.0       455.0         6.0777       <1H OCEAN  
13431      2231.0       742.0         4.1399          INLAND

In [121]:
housing_num.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4450     -118.21     34.07                42.0        902.0           318.0   
14253    -117.10     32.71                29.0       3422.0           713.0   
15338    -117.29     33.20                16.0       2150.0           461.0   
18156    -122.06     37.36                35.0       2693.0           493.0   
13431    -117.42     34.10                18.0       3977.0           809.0   

       population  households  median_income  
4450       1312.0       323.0         1.9375  
14253      2775.0       644.0         1.7075  
15338      1428.0       407.0         2.4754  
18156      1343.0       455.0         6.0777  
13431      2231.0       742.0         4.1399

# **NEW SECTION**

# **Pipeline. Sonlar**

## Biz malumotlarni deyarli tayyorlab bo'ldik lekin yuqoridagi barcha codlar biroz tushunarsiz yani ketma ket bir nechta usullardan foydalanganimiz uchun tushunish qiyin bo'lgan holga kelgan shu sababdan yangi drive ga boshidan tartib bilan yozamiz yani 03 darsga